lib/portlayer/storage/nfs/volume.go

```diff
@@ -0,0 +1,260 @@
+
+	target, err := mount.Mount(nfsTargetURL)
+	if err != nil {
+		return nil, err
+	}
+
+	defer mount.Unmount(target)
+
+	//we assume that nfsTargetURL.path already exists.
+	//make volumes directory
+	if _, err := target.MkDir(path.Join(nfsTargetURL.Path, VolumesDir), filePermissions); err != nil && !os.IsExist(err) {
+		return nil, err
+	}
+
+	//make metadata directory
+	if _, err := target.MkDir(path.Join(nfsTargetURL.Path, metadataDir), filePermissions); err != nil && !os.IsExist(err) {
+		return nil, err
+	}
+
+	selfLink, err := util.VolumeStoreNameToURL(storeName)
+	if err != nil {
+		return nil, err
+	}
+
+	v := &NFSv3VolumeStore{
+		Name:           storeName,
+		volumeStoreDir: nfsTargetURL.Path,
+		MH:             mount,
+		Target:         nfsTargetURL,
+		SelfLink:       selfLink,
+	}
+
+	return v, nil
+}
+
+// Returns the path to the vol relative to the given store.  The dir structure
+// for a vol in a nfs store is `<configured nfs server path>/volumes/<vol ID>/<volume contents>`.
+func (v *NFSv3VolumeStore) volDirPath(ID string) string {
+	return path.Join(v.volumeStoreDir, VolumesDir, ID)
+}
+
+func (v *NFSv3VolumeStore) volumesDir() string {
+	return path.Join(v.volumeStoreDir, VolumesDir)
+}
+
+// Returns the path to the metadata directory for a volume
+func (v *NFSv3VolumeStore) volMetadataDirPath(ID string) string {
+	return path.Join(v.volumeStoreDir, metadataDir, ID)
+}
+
+// Creates a volume directory and volume object for NFS based volumes
+func (v *NFSv3VolumeStore) VolumeCreate(op trace.Operation, ID string, store *url.URL, capacityKB uint64, info map[string][]byte) (*storage.Volume, error) {
+	target, err := v.MH.Mount(v.Target)
+	if err != nil {
+		return nil, err
+	}
+
+	defer v.MH.Unmount(target)
+
+	if _, err := target.MkDir(v.volDirPath(ID), filePermissions); err != nil {
+		return nil, err
+	}
+
+	backing := NewNFSVolumeDevice(v.Target, v.volDirPath(ID))
+
+	vol, err := storage.NewVolume(v.SelfLink, ID, info, backing)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := writeMetadata(op, v.volMetadataDirPath(ID), info, target); err != nil {
+		return nil, err
+	}
+
+	op.Infof("nfs volume (%s) successfully created on volume store (%s)", ID, v.Name)
+	return vol, nil
+}
+
+// Removes a volume and all of it's contents from the nfs store. We already know via the cache if it is in use.
+func (v *NFSv3VolumeStore) VolumeDestroy(op trace.Operation, vol *storage.Volume) error {
+	target, err := v.MH.Mount(v.Target)
+	if err != nil {
+		return err
+	}
+
+	defer v.MH.Unmount(target)
+
+	op.Infof("Attempting to remove volume (%s) and it's metadata from volume store (%s)", vol.ID, v.Name)
+
+	//remove volume directory and children
+	if err := target.RemoveAll(v.volDirPath(vol.ID)); err != nil {
+		op.Errorf("Failed to remove volume (%s) on volume store (%s) due to error (%s)", vol.ID, v.Name, err)
+		return err
+	}
+
+	//XXX: what should we do if we lose the nfs server between volume and metadata deletion?
```

Comment:
Yes I will repost these comments to a new tickets when we get the chance. My feeling is that we may possibly need to recover and mark the volume as deleted if we sanely can. The issue to me is that if we do not mark the volume as deleted the first time around we no longer actually have the volume. We just have some metadata left, so if the user gets confused or another(or the same) user try to attach that volume to container they are going to get a "does not exist" style error from us.